os/kernel/debug/mem_leak_checker.c

```diff
@@ -0,0 +1,327 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <stdlib.h>
+#include <debug.h>
+#include <stdio.h>
+#include <string.h>
+#include <queue.h>
+#include <sys/types.h>
+#include <tinyara/mm/mm.h>
+#include <tinyara/mm/heap_regioninfo.h>
+#include <binary_manager/binary_manager.h>
+
+#include "../binfmt/libelf/libelf.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define MEM_USED   0
+#define MEM_LEAK   1
+#define MEM_BROKEN 2
+
+#define MEM_ACCESS_UNIT    0x04
+#define MAX_ALLOC_COUNT    CONFIG_MEM_LEAK_CHECKER_MAX_ALLOC_COUNT
+#define HASH_SIZE          CONFIG_MEM_LEAK_CHECKER_HASH_TABLE_SIZE
+
+#define RAM_START_ADDR  kregionx_start[0]
+#define RAM_END_ADDR	(kregionx_start[CONFIG_KMM_REGIONS - 1] + kregionx_size[CONFIG_KMM_REGIONS - 1])
+#define HEAP_START_ADDR (void *)leak_checker.heap_start
+#define HEAP_END_ADDR   (void *)leak_checker.heap_end
+
+struct alloc_node_info_s {
+	volatile struct mm_allocnode_s *node;
+	struct alloc_node_info_s *next;
+};
+
+static struct alloc_node_info_s *hash_table[HASH_SIZE];
+static struct alloc_node_info_s node_info[MAX_ALLOC_COUNT];
+static struct mem_leak_checker_info_s leak_checker;
+
+#ifdef CONFIG_MM_KERNEL_HEAP
+extern struct mm_heap_s g_kmmheap[CONFIG_KMM_NHEAPS];
+#endif
+extern sq_queue_t g_bin_addr_list;
+static void hash_init(void)
+{
+	int index;
+	for (index = 0; index < HASH_SIZE; ++index) {
+		hash_table[index] = NULL;
+	}
+}
+
+static void add_hash(int index)
+{
+	long key;
+	struct alloc_node_info_s *cur;
+	key = (long)node_info[index].node % HASH_SIZE;
+	if (hash_table[key] == NULL) {
+		hash_table[key] = &node_info[index];
+		return;
+	}
+
+	cur = hash_table[key];
+	while (cur->next) {
+		cur = cur->next;
+	}
+	cur->next = &node_info[index];
+}
+
+static bool search_hash(unsigned long value)
+{
+	long key = value % HASH_SIZE;
+	struct alloc_node_info_s *cur = hash_table[key];
+	while (cur != NULL) {
+		if ((unsigned long)cur->node == value && (unsigned long)cur->node != (value + (unsigned long)SIZEOF_MM_ALLOCNODE)) {
+			if (cur->node->reserved == MEM_USED) {
+				return false;
+			}
+			cur->node->reserved = MEM_USED;
+			return true;
+		}
+		if (cur->next == NULL) {
+			return false;
+		}
+		cur = cur->next;
```

Comment:
Maybe you could modify the design of this loop?
The while condition in line 93 will not fail as you are checking the value of the next pointer separately in line 101,
unless of course the initialized value of cur itself is NULL